### PR TITLE
Sun Angles 

### PR DESCRIPTION
**Pull request by rs028**
PR to address issues #14  and #108. The equation for SZA is correct, but need to check the calculation of LHA and DEC.

Question 1: should variables `secx`, `cosx`, `lat `, `longt`, `lha`, `sinld`, `cosld` be arguments of the subroutine `zenith` now `calcZenith`? These are written to `photoRateCalcParameters.output` for diagnostic purposes

Question 2: are `latitude` and `longitude` global variables?

### CODE REVIEW

**Comment by rs028** [comment]
Wait to merge untile LHA and DEC calculations have been checked

File: utilityFunctions.f90 [hunk 1]
```diff
@@ -28,35 +32,75 @@ subroutine atmosphere( m, o2, n2 )
...
+    ! calculate cosine of solar zenith angle (cosx)
+    ! SZA is the angle between local vertical and center of the Sun
     sinld = sin( lat ) * sin( dec )
     cosld = cos( lat ) * cos( dec )
+    cosx = cos( lha ) * cosld + sinld
 
-    rampValue = ramp( cos( lha ) * cosld + sinld )
+    ! set negative cosx to zero and calculate the inverse (secx)
+    ! secx is needed for the MCM photolysis parameterisation
+    rampValue = ramp( cosx )
```
Comment (spco) [inline utilityFunctions.f90]: Could we perhaps rename `rampValue` to `ramped_cosx` or similar, to be more descriptive of what that value is?

File: utilityFunctions.f90 [hunk 2]
```diff
@@ -28,35 +32,75 @@ subroutine atmosphere( m, o2, n2 )
...
+    maxDecInRad = 23.44 * pi / 180.0
+    currentFYear = dayAsFractionOfYear + ( t / secondsInYear )
+    temporary = 2.0 * 0.0167 * sin( 2 * pi * ( currentFYear - ( 2.0 / daysInYear ) ) )
+    dec = asin( sin( -maxDecInRad ) * cos( ( 2.0 * pi ) * ( currentFYear + ( 10.0 / daysInYear ) ) + temporary ) )
+    return
+  end function calcDec  ! NB: is this in radians or degrees?
+
+  ! ----------------------------------------------------------------- !
+
+  subroutine calcZenith( ttime, dec )
```
Comment (spco) [inline utilityFunctions.f90]: Can we rename `ttime` to `t`, which fits the style of other functions?

File: utilityFunctions.f90 [hunk 3]
```diff
@@ -28,35 +32,75 @@ subroutine atmosphere( m, o2, n2 )
...
-    cosx = cos( lha ) * cosld + sinld
-
     return
-  end subroutine zenith
+  end subroutine calcZenith
+
+  ! ----------------------------------------------------------------- !
 
   pure function ramp( arg1 ) result ( rampValue )
+    ! if arg1 is positive returns arg1
```
Comment (spco) [inline utilityFunctions.f90]: While you're here, could you rewrite this function in a more transparent way? I'd prefer to see the function use `rampValue = max( 0, arg1 )` than the current formula (it gives the same result, but is much easier to understand!)

File: utilityFunctions.f90 (hunk 1 above)
Comment (rs028) [inline utilityFunctions.f90]: Actually, do we really need this as a separate function? I believe it is used only here. What if I just substitute lines 93-94 with:

    ramped_cosx = max(0, cosx )
    secx = 1.0 / ( rampValue + 1.00d-30 )

Comment (spco) [inline utilityFunctions.f90]: That sounds a good idea!

**Comment by rs028** [comment]
Okay, all done. What about my questions above?

**Comment by spco** [comment]
Ah, I forgot to answer those. (1) I think it's best to leave these as members of `zenithData`, rather than as parameters to pass in and out - it's easier and tidier to just pass the module, in my opinion.

(2) `latitude` and `longitude` are currently members of `SZACalcVars` module. I'm planning to combine those into `zenithData`, as that's fairly natural. I have that in a branch somewhere, I just haven't made a PR for that yet. I'll probably wait until everything else has been merged anyway, as that area of the code has various edits happening to it.

**Comment by rs028** [comment]
Okay sounds good,
I am thinking, if it makes sense to have latitude and longitude in degrees (as it is now) and redefine lat and longt in radians. At the moment only lat is in radians and longt is equal to longitude, which does not make sense.

**Comment by spco** [comment]
Perhaps the best way to fix this is to move the definition of `lat` after the `radians` definition, and on a single line each, e.g. `lat = latitude / radians` rather than its current use of multiple lines. Given that `longt` is not used in the formula, I'd actually just get rid of it.

On that point, surely `longitude` is needed for the `lha` calculation, given that the longitude will affect the local time? Or is the time data all given in the timezone associated to that longitude, not GMT?

**Comment by spco** [comment]
Let me know when you want me to review this again - for the record, here I think we need `max( 0.0, cosx )` as max can't compare an int against a real - both need to be real.

**Comment by rs028** [comment]
Okay, so I think if `lat` is just a subroutine variable, then we can get rid of `longt` and output `latitude` and `longitude` in degrees in `photoRateCalcParameters.output` which makes a lot more sense. Then merge `SZACalcVars` and `zenithData`, as you said. I can do it in this PR.

**Comment by spco** [comment]
Ok, feel free to do that merge of the modules. Outputting in degrees makes more sense, yes.

**Comment by rs028** [comment]
Should be all done, but wait to merge until I have checked the claculations of LHA and DEC.

Regarding longitude and LHA, it probably depends whether the model operates in GMT or Local Time, which I am not really sure to be honest. That's something to be checked. Maybe a separate issue?

**Comment by spco** [comment]
Conflicts due to #197 now resolved

**Comment by spco** [comment]
@rs028 what would you like to do here? I'm happy to fix these conflicts, merge this as it is, and have you submit another PR to check the LHA and DEC calculations, and check whether we're in local time or GMT? Would that work for you? It might be the simplest, and would give a clean slate for that checking.

**Comment by rs028** [comment]
Okay, now how do I avoid the mess I fell into while trying to rebase the `press_m` branch?

**Comment by spco** [comment]
@rs028 now this is merged, checking the LHA and DEC calculations and local/GMT time should be fairly straightforward :)

**Pull request merged**